File: ruby/jruby-netty/syslog-server/syslog-server.rb
#!/usr/bin/env ruby

require "java"
require File.join(File.dirname(__FILE__), "..", "netty-3.2.4.Final.jar")
java_import "com.loggly.syslog.SyslogDecoder"

class SyslogServerHandler < org.jboss.netty.channel.SimpleChannelHandler
  class << self
    include org.jboss.netty.channel.ChannelPipelineFactory
    def getPipeline
      return org.jboss.netty.channel.Channels.pipeline(SyslogDecoder.new, self.new)
    end # def getPipeline
  end # class << self 

  def initialize
    super
    @count = 0
    @start = Time.now
  end # def initialize

  def messageReceived(context, event)
    @count += 1
    if (@count % 100000 == 0)
      now = Time.now
      puts event.getMessage.toString
      rate = @count / (now - @start);
      puts("Rate: #{rate} (total: #{@count})");
    end
  end # def messageReceived

  def exceptionCaught(context, exception)
    exception.getCause.printStackTrace
    exception.getChannel.close
  end # def exceptionCaught
end # class SyslogServerHandler

class RubySyslogServer
  def initialize(host, port)
    @factory = org.jboss.netty.channel.socket.nio.NioServerSocketChannelFactory.new(
      java.util.concurrent.Executors.newCachedThreadPool(),
      java.util.concurrent.Executors.newCachedThreadPool()
    )

    @bootstrap = org.jboss.netty.bootstrap.ServerBootstrap.new(@factory)
    @bootstrap.setPipelineFactory(SyslogServerHandler)
    @bootstrap.setOption("child.tcpNoDelay", true);
    @bootstrap.setOption("child.keepAlive", true);

    @host = host
    @port = port
  end # def initialize

  def start
    address = java.net.InetSocketAddress.new(@host, @port)
    return @bootstrap.bind(address)
  end # def start

  def run
    future = start
  end # def run

  def self.main(args)
    host = args[0]
    port = args[1].to_i
    RubySyslogServer.new(host, port).run
  end # def self.main
end # class SyslogServer

if __FILE__ == $0
  RubySyslogServer.main(ARGV)
end

File: ruby/jruby-netty/syslog-server/com/loggly/syslog/SyslogServer.java
package com.loggly.syslog;

import java.net.InetSocketAddress;
import java.util.concurrent.Executors;
import org.jboss.netty.bootstrap.ServerBootstrap;
import java.nio.charset.Charset;
import org.jboss.netty.buffer.ChannelBuffer;
import org.jboss.netty.buffer.ChannelBufferIndexFinder;
import org.jboss.netty.buffer.ChannelBuffers;
import org.jboss.netty.channel.Channel;
import org.jboss.netty.channel.ChannelHandlerContext;
import org.jboss.netty.channel.ChannelPipeline;
import org.jboss.netty.channel.ChannelPipelineFactory;
import org.jboss.netty.channel.Channels;
import org.jboss.netty.channel.ChannelStateEvent;
import org.jboss.netty.channel.ExceptionEvent;
import org.jboss.netty.channel.MessageEvent;
import org.jboss.netty.channel.SimpleChannelHandler;
import org.jboss.netty.channel.socket.nio.NioServerSocketChannelFactory;
import org.jboss.netty.handler.codec.replay.ReplayingDecoder;
import com.loggly.syslog.SyslogDecoder;

public class SyslogServer {
  public class SyslogHandler extends SimpleChannelHandler {
    public int count;
    public long start;

    public SyslogHandler() {
      super();
      this.count = 0;
      this.start = System.currentTimeMillis();
    }

    public void messageReceived(ChannelHandlerContext context, MessageEvent event) {
      count += 1;
      if (count % 100000 == 0) {
        //ChannelBuffer buffer = (ChannelBuffer) event.getMessage();
        //System.out.println(buffer.toString(Charset.defaultCharset()));
        System.out.println(event.getMessage());
        double rate = (double)this.count / (((double)System.currentTimeMillis() - this.start) / 1000.);
        System.out.println("Rate: " + rate + " (total: " + this.count);
      }
    }

    public void exceptionCaught(ChannelHandlerContext context, ExceptionEvent event) {
      event.getCause().printStackTrace();
      event.getChannel().close();
    }
  } /* class SyslogHandler */

  private NioServerSocketChannelFactory factory;
  private ServerBootstrap bootstrap;
  private int port;
  private String host;

  public SyslogServer(String host, int port) {
    this.host = host;
    this.port = port;
    this.factory = new NioServerSocketChannelFactory(
      Executors.newCachedThreadPool(), 
      Executors.newCachedThreadPool()
    );

    this.bootstrap = new ServerBootstrap(factory);
    this.bootstrap.setPipelineFactory(new ChannelPipelineFactory() {
      public ChannelPipeline getPipeline() {
        return Channels.pipeline(new SyslogDecoder(), new SyslogHandler());
      }
    });
  }

  public void start() {
    this.bootstrap.bind(new InetSocketAddress(this.host, this.port));
  }

  public static void main(String[] args) {
    String host;
    int port;
    host = args[0];
    port = Integer.parseInt(args[1]);
    SyslogServer server = new SyslogServer(host, port);
    server.start();
  }
}
